File: OJ/leet1122/main.cpp
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

class Solution {
public:
    vector<int> relativeSortArray(vector<int>& arr1, vector<int>& arr2) {
        vector<int> index(1005, -1);
        vector<int> cnt(arr2.size(), 0), rest;
        vector<int> ans(arr1.size(), 0);

        for (int i=0; i<arr2.size(); ++i) index[arr2[i]] = i;
        // printf("\n arr1: "); for (auto x: arr1) printf("%d, ", x);
        // printf("\n index: "); for (auto x: index) printf("%3d, ", x);

        for (auto ai:arr1) {
            if (index[ai] == -1) {
                rest.push_back(ai);
            }
            else {
                cnt[index[ai]] += 1;
            }
        }

        sort(rest.begin(), rest.end());
        // printf("\n cnt:   "); for (auto x: cnt) printf("%3d, ", x);
        // printf("\n rest:   "); for (auto x: rest) printf("%3d, ", x);
        // printf("\n");

        int start = 0;
        for (int i=0; i<cnt.size(); ++i)  {
            int ci = cnt[i];
            while (ci--) ans[start++] = arr2[i];
        }
        for (auto ri: rest) {
            ans[start++] = ri;
        }
        return ans;
    }
};

int main(int argc, char const *argv[])
{
    Solution sol;
    vector<int> arr1 = {2, 3, 1, 3, 2, 4, 6, 7, 9, 2, 19},  
                arr2 = {2, 1, 4, 3, 9, 6};
    vector<int> ans = sol.relativeSortArray(arr1, arr2);
    for (auto ai: ans) printf("%d, ", ai);
    return 0;
}

File: OJ/leet1122/src/main.rs
struct Solution;

impl Solution {
    pub fn relative_sort_array(arr1: Vec<i32>, arr2: Vec<i32>) -> Vec<i32> {
        let mut index = vec![9999; 1005];
        let mut cnt = vec![0; arr2.len()];
        for (i, ai) in arr2.iter().enumerate() {
            index[*ai as usize] = i;
        }
        
        let mut rest = Vec::new();
        for ai in arr1.into_iter() {
            if index[ai as usize] == 9999 {
                rest.push(ai);
            }
            else {
                cnt[index[ai as usize]] += 1;
            }
        }

        rest.sort();

        let mut ans = Vec::new();
        for (i, ci) in cnt.into_iter().enumerate() {
            for _ in 0..ci {
                ans.push(arr2[i]);
            }
        }
        for ri in rest.into_iter() {
            ans.push(ri);
        }
        ans
    }
}

fn main() {
    let arr1 = vec![2,3,1,3,2,4,6,7,9,2,19];
    let arr2 = vec![2,1,4,3,9,6];
    println!(" = {:?}", Solution::relative_sort_array(arr1, arr2));
}